### ISSUE TITLE
Unable to assign azurerm_data_factory managed identity to a group

### ISSUE THREAD

**Issue by rob-dolman-foureyesinsight** [hashicorp/terraform-provider-azurerm#15027]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v1.1.3
on windows_amd64
+ provider registry.terraform.io/hashicorp/azurerm v2.92.0
```

### Affected Resource(s)

* `azurerm_data_factory`

### Terraform Configuration Files

```hcl
resource "azurerm_data_factory" "data_factory" {
  name                = var.name
  location            = var.location
  resource_group_name = var.resource_group_name

  identity {
    type = "SystemAssigned"
  }

  lifecycle {
    ignore_changes = [
      global_parameter,
      vsts_configuration
    ]
  }
}

data "azuread_group" "group" {
  display_name   = "My-DataFactory-Group"
}

resource "azuread_group_member" "assign_access" {
  group_object_id  = data.azuread_group.group.id
  member_object_id = azurerm_data_factory.data_factory.identity[0].principal_id
}
```

### Expected Behaviour

A `azurerm_data_factory` resource should be created and it's managed identity assigned to the group `My-DataFactory-Group`

### Actual Behaviour

A `azurerm_data_factory` resource is created but it's managed identity is not immediately available:

```
Error: Could not retrieve principal object "<MANAGED IDENTITY ID>"
DirectoryObjects.BaseClient.Get(): unexpected status 403 with OData error: Authorization_RequestDenied: Insufficient privileges to complete the operation.
```

### Steps to Reproduce

1. Create an azure group `My-DataFactory-Group`
2. Assign the terraform security context (user / principal) as an owner of the group
2. Run `terraform apply` with the above configuration

### Workaround

Adding an artificial 30 second wait to the group assignment seems to solve the problem (is there perhaps some AD propagation that needs to occur?)

```
resource "time_sleep" "data_factory_wait" {
  depends_on = [azurerm_data_factory.data_factory]
  create_duration = "30s"
}

resource "azuread_group_member" "assign_access" {
  depends_on       = [time_sleep.data_factory_wait]
  group_object_id  = data.azuread_group.group.id
  member_object_id = azurerm_data_factory.data_factory.identity[0].principal_id
}
```

**Comment by jackofallops** [comment]
Hi @rob-dolman-foureyesinsight - At first glance, this looks to be a permissions issue on the `azuread` provider user/principal being used. i.e. it has insufficient AAD permissions to read that resource. Can you check the user/principal being used has the appropriate [API permissions](https://registry.terraform.io/providers/hashicorp/azuread/latest/docs/resources/group_member#api-permissions)?

**Comment by rob-dolman-foureyesinsight** [comment]
Hi @jackofallops I assumed the same at first, but check repro step #2 and the workaround of waiting 30 seconds. The 403 isn't returned because it has insufficient AAD permissions to read the group (which it is an owner of), it somehow has insufficient permissions to read the data factory managed identity (that's just been created), however, after 30 seconds it has no problem with permissions

**Comment by jackofallops** [comment]
Hi @rob-dolman-foureyesinsight - This does sound like an eventual consistency / replication issue for the identity then. We're beholden to Azure's replication timings in this, and we do make reasonable attempts to ensure replication is complete, but ultimately we cannot guarantee it. You're already using the mechanism I'd normally suggest to compensate in the `time` provider, so I can't offer any additional workarounds currently. It may be that there are additional steps that can be taken in the `azurerm_data_factory ` resource to help ensure the principal is ready, but as I've mentioned, I suspect there would still be no guarantee that the ad_group_member resource call would succeed.

**Comment by rob-dolman-foureyesinsight** [comment]
Yeah fair enough, I suspect there is something unique about the way in which data factory creates the managed identity for itself since other resources with managed identities (e.g. AKS) don't seem to suffer from this problem. The workaround isn't painful at all so hopefully this issue can help others
